File: app/src/main/java/ru/samtakoy/core/presentation/ScheduleEditFragment.java
package ru.samtakoy.core.presentation;

import android.app.Activity;
import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.DialogFragment;

import ru.samtakoy.R;
import ru.samtakoy.core.database.room.entities.elements.Schedule;
import ru.samtakoy.core.database.room.entities.elements.ScheduleItem;
import ru.samtakoy.core.database.room.entities.elements.ScheduleTimeUnit;

public class ScheduleEditFragment extends DialogFragment {

    public static final String TAG = "ScheduleEditFragment";

    private static final String ARG_SCHEDULE_STRING = "ARG_SCHEDULE_STRING";

    private static final String SAVED_SCHEDULE_CUR_ITEM_TO_ADD = "SAVED_SCHEDULE_CUR_ITEM_TO_ADD";

    public static final String RESULT_SCHEDULE_STRING = "RESULT_SCHEDULE_STRING";

    public static ScheduleEditFragment newFragment(Schedule schedule){
        ScheduleEditFragment result = new ScheduleEditFragment();
        Bundle args = new Bundle();
        args.putString(ARG_SCHEDULE_STRING, schedule == null ? null : schedule.toString());
        result.setArguments(args);
        return result;
    }

    private Schedule mSchedule_;
    private ScheduleItem mScheduleCurItemToAdd;

    private TextView mScheduleView;
    private Button mClearScheduleButton;

    private Button mScheduleCurItemButton;
    private Button mScheduleCurItemClearButton;
    private Button mScheduleCurItemAddButton;
    private ViewGroup mScheduleItemButtons;

    /*public interface Callbacks{
        void onScheduleSelect(Schedule schedule);
    }/**/

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        readArgs(getArguments());
        if(savedInstanceState != null){
            String strScheduleItem = savedInstanceState.getString(SAVED_SCHEDULE_CUR_ITEM_TO_ADD, null);
            //if(strScheduleItem != null) {
                mScheduleCurItemToAdd = ScheduleItem.fromString(strScheduleItem);
            //}
        }
        if(mScheduleCurItemToAdd == null){
            mScheduleCurItemToAdd = new ScheduleItem(1, ScheduleTimeUnit.MINUTE);
        }
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        //outState.putString(SAVED_SCHEDULE_CUR_ITEM_TO_ADD, mScheduleCurItemToAdd == null ? null : mScheduleCurItemToAdd.toString());
        outState.putString(SAVED_SCHEDULE_CUR_ITEM_TO_ADD, mScheduleCurItemToAdd.toString());
    }

    @Override
    public void onDestroy() {
        mSchedule_ = null;
        super.onDestroy();
    }

    private void readArgs(Bundle arguments) {
        mSchedule_ = new Schedule();
        String strSchedule = arguments.getString(ARG_SCHEDULE_STRING);
        if(strSchedule != null){
            Log.e(TAG, "readArgs:"+strSchedule);
            mSchedule_.initFromString(strSchedule);
        }
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {

        LayoutInflater inflater = LayoutInflater.from(getActivity());
        View v = inflater.inflate(R.layout.fragment_schedule_edit, null);

        initView(v);

        return new AlertDialog.Builder(getActivity())
                .setView(v)
                .setCancelable(true)
                .setPositiveButton(R.string.btn_ok, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int i) {
                        trySendResult();
                    }
                })
                .create();
    }

    private void initView(View v) {
        mScheduleView = v.findViewById(R.id.schedule_text);

        mClearScheduleButton = v.findViewById(R.id.clear_btn);
        mClearScheduleButton.setOnClickListener((view)->clearSchedule());

        mScheduleCurItemButton = v.findViewById(R.id.item_btn);

        mScheduleCurItemClearButton = v.findViewById(R.id.item_clear_btn);
        mScheduleCurItemClearButton.setOnClickListener((view)->clearScheduleCurItem());

        mScheduleCurItemAddButton = v.findViewById(R.id.item_add_btn);
        mScheduleCurItemAddButton.setOnClickListener((view)->addScheduleCurItem());

        mScheduleItemButtons = v.findViewById(R.id.schedule_item_btns);

        createScheduleItemButtons(v.findViewById(R.id.schedule_item_btns));
        updateScheduleView();
        updateScheduleItemView();
    }


    @Override
    public void onDestroyView() {

        mScheduleView = null;
        mClearScheduleButton = null;
        mScheduleCurItemButton = null;
        mScheduleCurItemClearButton = null;
        mScheduleCurItemAddButton = null;
        mScheduleItemButtons = null;

        super.onDestroyView();
    }

    private void updateScheduleView(){
        if(mSchedule_.isEmpty()){
            mScheduleView.setText(R.string.schedule_none);
        } else {
            mScheduleView.setText(mSchedule_.toStringView(getResources()));
        }
    }

    private void updateScheduleItemView(){
        mScheduleCurItemButton.setText(
                mScheduleCurItemToAdd.toStringView(getResources())
        );
    }

    private void createScheduleItemButtons(ViewGroup container) {

        Context ctx = getActivity();
        for(ScheduleTimeUnit unit:ScheduleTimeUnit.values()) {
            Button unitBtn = new Button(ctx);
            unitBtn.setText("+1"+getResources().getString(unit.getStringsId()));
            unitBtn.setOnClickListener((View view)->onScheduleTimeUnitSelect(unit));
            container.addView(unitBtn);
        }
    }

    private void clearSchedule() {
        mSchedule_.clear();
        updateScheduleView();
    }

    private void clearScheduleCurItem() {
        mScheduleCurItemToAdd.setValue(1, ScheduleTimeUnit.MINUTE);
        updateScheduleItemView();
    }

    private void addScheduleCurItem() {
        mSchedule_.addItem(mScheduleCurItemToAdd.copy());
        updateScheduleView();
    }

    private void onScheduleTimeUnitSelect(ScheduleTimeUnit unit) {
        if(unit == mScheduleCurItemToAdd.getTimeUnit()){
            mScheduleCurItemToAdd.setValue(1+mScheduleCurItemToAdd.getDimension(), unit);
        } else{
            mScheduleCurItemToAdd.setValue(1, unit);
        }
        updateScheduleItemView();
    }

    private void trySendResult() {

        if(getTargetFragment() == null){
            Log.e(TAG, "Target fragment is missing");
            return;
        }

        Intent result = new Intent();
        result.putExtra(RESULT_SCHEDULE_STRING, mSchedule_.toString());
        getTargetFragment().onActivityResult(getTargetRequestCode(), Activity.RESULT_OK, result);
    }

}

File: app/src/main/kotlin/ru/samtakoy/core/presentation/schedule/ScheduleEditFragment.kt
package ru.samtakoy.core.presentation.schedule

import android.app.Activity
import android.app.Dialog
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.core.os.bundleOf
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.viewModels
import ru.samtakoy.R
import ru.samtakoy.core.app.di.Di
import ru.samtakoy.core.presentation.base.observe
import ru.samtakoy.core.presentation.base.viewmodel.AbstractViewModel
import ru.samtakoy.core.presentation.base.viewmodel.ViewModelOwner
import ru.samtakoy.core.presentation.schedule.vm.ScheduleEditViewModel
import ru.samtakoy.core.presentation.schedule.vm.ScheduleEditViewModel.Event
import ru.samtakoy.core.presentation.schedule.vm.ScheduleEditViewModelFactory
import ru.samtakoy.core.presentation.schedule.vm.ScheduleEditViewModelImpl
import ru.samtakoy.features.learncourse.domain.model.schedule.ScheduleTimeUnit
import ru.samtakoy.features.learncourse.domain.model.schedule.serialize.ParcelableSchedule
import ru.samtakoy.features.learncourse.domain.model.schedule.serialize.toDomainOrEmpty
import javax.inject.Inject

class ScheduleEditFragment : DialogFragment(), ViewModelOwner {
    private var mScheduleView: TextView? = null
    private var mClearScheduleButton: Button? = null

    private var mScheduleCurItemButton: Button? = null
    private var mScheduleCurItemClearButton: Button? = null
    private var mScheduleCurItemAddButton: Button? = null
    private var mScheduleItemButtons: ViewGroup? = null

    @Inject
    internal lateinit var viewModelFactory: ScheduleEditViewModelFactory.Factory
    private val viewModel by viewModels<ScheduleEditViewModelImpl> {
        viewModelFactory.create(
            schedule = (requireArguments().getParcelable(ARG_SCHEDULE_STRING) as? ParcelableSchedule).toDomainOrEmpty()
        )
    }
    override fun getViewModel(): AbstractViewModel = viewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        Di.appComponent.inject(this)
        super.onCreate(savedInstanceState)
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val v = layoutInflater.inflate(R.layout.fragment_schedule_edit, null)

        initView(v)

        return AlertDialog.Builder(requireActivity())
            .setView(v)
            .setCancelable(true)
            .setPositiveButton(R.string.btn_ok, object : DialogInterface.OnClickListener {
                override fun onClick(dialogInterface: DialogInterface?, i: Int) {
                    viewModel.onEvent(Event.ConfirmResultClick)
                }
            })
            .create()
    }

    override fun onObserveViewModel() {
        super.onObserveViewModel()
        viewModel.getViewActionsAsFlow().observe(viewLifecycleOwner, ::onAction)
        viewModel.getViewStateAsFlow().observe(viewLifecycleOwner, ::onState)
    }

    private fun onState(state: ScheduleEditViewModel.State) {
        mScheduleView!!.setText(state.schedule)
        mScheduleCurItemButton!!.setText(state.scheduleCurItemButtonText)
    }

    private fun onAction(action: ScheduleEditViewModel.Action) {
        when (action) {
            is ScheduleEditViewModel.Action.CloseWithResult -> {
                trySendResult(action.serializedSchedule)
            }
        }
    }

    private fun initView(v: View) {
        mScheduleView = v.findViewById<TextView?>(R.id.schedule_text)

        mClearScheduleButton = v.findViewById<Button?>(R.id.clear_btn)
        mClearScheduleButton!!.setOnClickListener(
            View.OnClickListener { view: View? ->
                viewModel.onEvent(Event.ClearScheduleClick)
            }
        )

        mScheduleCurItemButton = v.findViewById<Button?>(R.id.item_btn)

        mScheduleCurItemClearButton = v.findViewById<Button?>(R.id.item_clear_btn)
        mScheduleCurItemClearButton!!.setOnClickListener(
            View.OnClickListener { view: View? ->
                viewModel.onEvent(Event.ClearScheduleCurItemClick)
            }
        )

        mScheduleCurItemAddButton = v.findViewById<Button?>(R.id.item_add_btn)
        mScheduleCurItemAddButton!!.setOnClickListener(
            View.OnClickListener { view: View? ->
                viewModel.onEvent(Event.AddScheduleCurItem)
            }
        )

        mScheduleItemButtons = v.findViewById<ViewGroup?>(R.id.schedule_item_btns)

        createScheduleItemButtons(v.findViewById<ViewGroup?>(R.id.schedule_item_btns))
    }

    override fun onDestroyView() {
        mScheduleView = null
        mClearScheduleButton = null
        mScheduleCurItemButton = null
        mScheduleCurItemClearButton = null
        mScheduleCurItemAddButton = null
        mScheduleItemButtons = null

        super.onDestroyView()
    }

    private fun createScheduleItemButtons(container: ViewGroup) {
        val ctx: Context? = getActivity()
        if (container.childCount < ScheduleTimeUnit.entries.size) {
            for (unit in ScheduleTimeUnit.entries) {
                val unitBtn = Button(ctx)
                unitBtn.setText("+1" + getResources().getString(unit.textStringId))
                unitBtn.setOnClickListener(
                    View.OnClickListener { view: View? ->
                        viewModel.onEvent(Event.ScheduleTimeUnitSelect(unit))
                    }
                )
                container.addView(unitBtn)
            }
        }
    }

    private fun trySendResult(serializedSchedule: ParcelableSchedule) {
        if (getTargetFragment() == null) {
            Log.e(TAG, "Target fragment is missing")
            return
        }

        val result = Intent()
        result.putExtra(RESULT_SCHEDULE_STRING, serializedSchedule)
        getTargetFragment()!!.onActivityResult(getTargetRequestCode(), Activity.RESULT_OK, result)
    }

    companion object {
        const val TAG: String = "ScheduleEditFragment"

        private const val ARG_SCHEDULE_STRING = "ARG_SCHEDULE_STRING"
        const val RESULT_SCHEDULE_STRING: String = "RESULT_SCHEDULE_STRING"

        @JvmStatic fun newFragment(schedule: ParcelableSchedule?): ScheduleEditFragment {
            return ScheduleEditFragment().apply {
                arguments = bundleOf(
                    ARG_SCHEDULE_STRING to if (schedule == null) {
                        null
                    } else {
                        schedule
                    }
                )
            }
        }
    }
}
